engine/storage/snapshot/src/org/apache/cloudstack/storage/snapshot/SnapshotDataFactoryImpl.java

```diff
@@ -64,6 +64,24 @@ public SnapshotInfo getSnapshot(DataObject obj, DataStore store) {
         return so;
     }
 
+    @Override
+    public List<SnapshotInfo> getSnapshots(long volumeId, DataStoreRole role) {
+
+        SnapshotDataStoreVO snapshotStore = snapshotStoreDao.findByVolume(volumeId, role);
+        if (snapshotStore == null) {
+            return null;
```

Comment:
This will cause a NPE. Consider returning an empty list.

Reply:
@rafaelweingartner  Deleting snapshots from primary after copy is complete taken care in xenserver case in postSnapshotCreation method in XenserverSnapshotStrategy.java for all snapshots except the lost one. So this one only takes care of the last snapshot that we keep in primary for delta snapshots.  null return is replaced with empty set

Reply:
@rafaelweingartner  the problem related to SR_BACKEND_FAILURE_109 does not seem to be related to this